src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -0,0 +1,206 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.Eventing;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WebJobs.Script.WebHost.Diagnostics
+{
+    public static class MetricsEventManager
+    {        
+        private static FunctionActivityTracker instance = null;
+
+        private static object functionActivityTrackerLockObject = new object();
+
+        public static void BeginFunctionEventHandler()
+        {
+            lock (functionActivityTrackerLockObject)
```

Comment:
These calls are on the hot path - called for every function invocation. How can we remove all this locking?